core/src/appleMain/kotlin/Scanner.kt

```diff
@@ -16,33 +18,78 @@ import platform.CoreBluetooth.CBManagerStateUnsupported
 
 public class AppleScanner internal constructor(
     central: CentralManager,
-    services: List<Uuid>?,
+    filters: List<Filter>,
     options: Map<Any?, *>?,
     logging: Logging,
 ) : Scanner {
 
+    init {
+        require(filters.none { it is Filter.Address }) {
+            "Filtering by address (`Filter.Address`) is not supported on Apple platforms"
+        }
+    }
+
+    private val logger = Logger(logging, tag = "Kable/Scanner", identifier = null)
+
+    private val serviceFilters = filters.filterIsInstance<Filter.Service>().map(Filter.Service::uuid)
+
+    // Native filtering of advertisements is performed if: `filters` contains only (and at least one) `Filter.Service`.
+    private val nativeServiceFiltering = filters.isNotEmpty() && filters.all { it is Filter.Service }
+
+    init {
+        if (!nativeServiceFiltering) {
+            logger.warn {
+                message = "According to Core Bluetooth documentation: " +
+                    "\"The recommended practice is to populate the serviceUUIDs parameter rather than leaving it nil.\" " +
+                    "This means providing only (and at least 1) filter(s) of type `Filter.Service` to Scanner. " +
+                    "See https://developer.apple.com/documentation/corebluetooth/cbcentralmanager/1518986-scanforperipheralswithservices#discussion for more details."
+            }
+        }
+    }
+
     public override val advertisements: Flow<Advertisement> =
         central.delegate
             .response
             .onStart {
                 central.awaitPoweredOn()
-                central.scanForPeripheralsWithServices(services, options = options)
+                if (nativeServiceFiltering) {
+                    logger.info { message = "Starting scan with native service filtering" }
+                    central.scanForPeripheralsWithServices(serviceFilters, options)
+                } else {
+                    logger.info { message = "Starting scan with non-native filtering" }
+                    central.scanForPeripheralsWithServices(null, options)
+                }
             }
             .onCompletion {
+                logger.info { message = "Stopping scan" }
                 central.stopScan()
             }
             .filterIsInstance<DidDiscoverPeripheral>()
+            .filter { didDiscoverPeripheral ->
+                if (nativeServiceFiltering) return@filter true // Short-circuit (i.e. don't filter) when scan is using native service filtering.
```

Comment:
Why is this short-circuiting done? Maybe I dont understand "native service filtering"

Reply:
Native service filtering means that the OS is performing the filtering, so we short-circuit (return `true`) in flow `filter` operator so that we let all the events through the flow (since the OS will have already done the filtering for us).